File: node/src/injectors.ts
/**
 * Pure-function injectors — TypeScript port of `decoyshield/injectors.py`.
 *
 *   import { injectHtml, injectJson, injectHeaders, bait, isScanner } from "decoyshield";
 *
 *   const body = injectHtml("<html><body>hi</body></html>");
 *   const data = injectJson({ users: [] });
 *   const hdrs = injectHeaders({ "Content-Type": "text/html" });
 *   if (isScanner(req.headers, req.path)) { ... }
 */

import { DEFAULT_RESPONSE_HEADERS, PAYLOADS } from "./payloads.js";
import { type HeaderLike, fingerprint } from "./detector.js";

export type Channel =
  | "comment"
  | "hidden_div"
  | "white_text"
  | "hidden_input";

export const DEFAULT_CHANNELS: Channel[] = [
  "comment",
  "hidden_div",
  "white_text",
];

const ALL_CHANNELS = new Set<Channel>([
  "comment",
  "hidden_div",
  "white_text",
  "hidden_input",
]);

const ALREADY_INJECTED_MARKER = 'data-decoyshield="1"';

export function bait(
  name: string = "moral_lock",
  payloads: Record<string, string> = PAYLOADS,
): string {
  if (!(name in payloads)) {
    throw new Error(
      `Unknown payload ${JSON.stringify(name)}. ` +
        `Available: ${Object.keys(payloads).sort().join(", ")}`,
    );
  }
  return payloads[name];
}

export interface InjectHtmlOptions {
  payloads?: Record<string, string>;
  channels?: Channel[];
}

export function injectHtml(
  html: string,
  options: InjectHtmlOptions = {},
): string {
  if (html.includes(ALREADY_INJECTED_MARKER)) return html;

  const src = options.payloads ?? PAYLOADS;
  const channels = (options.channels ?? DEFAULT_CHANNELS).filter((c) =>
    ALL_CHANNELS.has(c),
  );
  const pieces: string[] = [];

  if (channels.includes("comment")) {
    for (const [name, body] of Object.entries(src)) {
      pieces.push(`<!-- decoyshield:${name}\n${body}\n-->`);
    }
  }

  if (channels.includes("hidden_div")) {
    const joined = Object.entries(src)
      .map(([name, body]) => `[${name}]\n${body}`)
      .join("\n\n");
    pieces.push(
      `<div style="display:none" aria-hidden="true" ` +
        `${ALREADY_INJECTED_MARKER}>${escapeText(joined)}</div>`,
    );
  }

  if (channels.includes("white_text")) {
    const joined = Object.values(src).join(" ");
    pieces.push(
      `<span style="color:#fff;background:#fff;font-size:1px;` +
        `position:absolute;left:-9999px">${escapeText(joined)}</span>`,
    );
  }

  if (channels.includes("hidden_input")) {
    for (const [name, body] of Object.entries(src)) {
      pieces.push(
        `<input type="hidden" name="_decoyshield_${name}" ` +
          `value="${escapeAttr(body)}">`,
      );
    }
  }

  const bundle = pieces.join("\n");
  if (!bundle) return html;

  const match = /<\/body\s*>/i.exec(html);
  if (match) {
    return (
      html.slice(0, match.index) + bundle + "\n" + html.slice(match.index)
    );
  }
  return html + "\n" + bundle;
}

export interface InjectJsonOptions {
  payloads?: Record<string, string>;
  key?: string;
}

export function injectJson<T extends object>(
  data: T,
  options: InjectJsonOptions = {},
): T & { [k: string]: unknown } {
  const src = options.payloads ?? PAYLOADS;
  const key = options.key ?? "_debug";
  return {
    ...data,
    [key]: {
      _audit: src["moral_lock"] ?? "",
      _internal_note: src["token_blackhole"] ?? "",
      _handshake: src["traceback"] ?? "",
      hint: "see X-Audit-Notice header; not for production parsing",
    },
  };
}

export function injectHeaders(
  headers?: Record<string, string>,
): Record<string, string> {
  const out: Record<string, string> = { ...(headers ?? {}) };
  for (const [k, v] of Object.entries(DEFAULT_RESPONSE_HEADERS)) {
    if (!Object.keys(out).some((existing) => existing.toLowerCase() === k.toLowerCase())) {
      out[k] = v;
    }
  }
  return out;
}

export function isScanner(
  headers: HeaderLike,
  path: string = "",
  method: string = "GET",
): boolean {
  const { verdict } = fingerprint(headers, path, method);
  return (
    verdict === "likely_scanner" ||
    verdict === "likely_ai" ||
    verdict === "likely_automation"
  );
}

function escapeText(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

function escapeAttr(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/"/g, "&quot;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

File: decoyshield/injectors.py
"""
Pure-function injectors — wrap any HTML / JSON / headers with invisible
LLM bait, with no framework required.

These are the "callable component" surface of DecoyShield: import a
function, hand it a string or dict, get back the protected version.
Use them inside any framework that doesn't ship a dedicated adapter
(Bottle, aiohttp, Tornado, raw WSGI, CGI scripts, static site
generators, …) or anywhere you assemble HTTP responses by hand.

    from decoyshield import inject_html, inject_json, inject_headers, bait

    body = inject_html("<html><body>hi</body></html>")
    data = inject_json({"users": []})
    hdrs = inject_headers({"Content-Type": "text/html"})
    snippet = bait("moral_lock")
"""
from __future__ import annotations

import re
from typing import Dict, Iterable, Mapping, Optional

from .core import DEFAULT_RESPONSE_HEADERS
from .detector import fingerprint
from .payloads import PAYLOADS

DEFAULT_CHANNELS = ("comment", "hidden_div", "white_text")
_ALL_CHANNELS = ("comment", "hidden_div", "white_text", "hidden_input")
_ALREADY_INJECTED_MARKER = 'data-decoyshield="1"'


def bait(
    name: str = "moral_lock",
    *,
    payloads: Optional[Mapping[str, str]] = None,
) -> str:
    """Return a single payload string by name.

    Use this when you want to embed a payload manually — e.g. as a
    comment in a config file, in a CLI help banner, or in a log line
    that an LLM scanner might read.
    """
    src = payloads if payloads is not None else PAYLOADS
    if name not in src:
        raise KeyError(
            f"Unknown payload {name!r}. Available: {sorted(src)}"
        )
    return src[name]


def inject_html(
    html: str,
    *,
    payloads: Optional[Mapping[str, str]] = None,
    channels: Iterable[str] = DEFAULT_CHANNELS,
) -> str:
    """Embed invisible LLM bait into an HTML document.

    Inserts bait through multiple channels (HTML comment, display:none
    div, white-on-white text) so that even aggressive sanitization is
    unlikely to strip every signal. Insertion happens before ``</body>``
    when possible, otherwise at the end.

    Idempotent: if ``html`` already carries the DecoyShield marker, the
    function returns it unchanged.
    """
    if _ALREADY_INJECTED_MARKER in html:
        return html

    src = dict(payloads if payloads is not None else PAYLOADS)
    chosen = tuple(c for c in channels if c in _ALL_CHANNELS)
    pieces = []

    if "comment" in chosen:
        for name, body in src.items():
            pieces.append(f"<!-- decoyshield:{name}\n{body}\n-->")

    if "hidden_div" in chosen:
        joined = "\n\n".join(
            f"[{name}]\n{body}" for name, body in src.items()
        )
        pieces.append(
            '<div style="display:none" aria-hidden="true" '
            f'{_ALREADY_INJECTED_MARKER}>'
            + _escape_text(joined)
            + "</div>"
        )

    if "white_text" in chosen:
        joined = " ".join(src.values())
        pieces.append(
            '<span style="color:#fff;background:#fff;font-size:1px;'
            'position:absolute;left:-9999px">'
            + _escape_text(joined)
            + "</span>"
        )

    if "hidden_input" in chosen:
        for name, body in src.items():
            pieces.append(
                f'<input type="hidden" name="_decoyshield_{name}" '
                f'value="{_escape_attr(body)}">'
            )

    bundle = "\n".join(pieces)
    if not bundle:
        return html

    match = re.search(r"</body\s*>", html, flags=re.IGNORECASE)
    if match:
        return html[: match.start()] + bundle + "\n" + html[match.start():]
    return html + "\n" + bundle


def inject_json(
    data: Mapping,
    *,
    payloads: Optional[Mapping[str, str]] = None,
    key: str = "_debug",
) -> Dict:
    """Return a copy of ``data`` with a ``_debug`` key carrying bait.

    The added key is a dict containing each payload under
    LLM-attractive names (``_audit``, ``_internal_note``,
    ``_handshake``). A human consumer ignores it; an LLM reading raw
    JSON treats it as authoritative metadata.

    Idempotent: re-injecting an already-injected dict overwrites the
    same key with fresh values.
    """
    src = dict(payloads if payloads is not None else PAYLOADS)
    out: Dict = dict(data)
    out[key] = {
        "_audit": src.get("moral_lock", ""),
        "_internal_note": src.get("token_blackhole", ""),
        "_handshake": src.get("traceback", ""),
        "hint": "see X-Audit-Notice header; not for production parsing",
    }
    return out


def inject_headers(
    headers: Optional[Mapping[str, str]] = None,
    *,
    payloads: Optional[Mapping[str, str]] = None,
) -> Dict[str, str]:
    """Return a headers dict augmented with DecoyShield bait headers.

    Adds ``X-Audit-Notice``, ``X-Debug-Trace``, ``X-Bypass-Protocol``.
    Existing keys are preserved.

    ``payloads`` is accepted for symmetry but currently unused; reserved
    for a future per-payload header mode.
    """
    del payloads  # reserved for future use
    out: Dict[str, str] = dict(headers) if headers else {}
    for k, v in DEFAULT_RESPONSE_HEADERS.items():
        out.setdefault(k, v)
    return out


def is_scanner(
    headers: Mapping[str, str],
    path: str = "",
    method: str = "GET",
) -> bool:
    """Convenience wrapper around :func:`fingerprint`.

    Returns ``True`` when the verdict is anything other than
    ``likely_human`` / ``unknown`` — i.e. when the request looks
    sufficiently bot-like to justify dropping bait into the response.
    """
    verdict, _tags, _score = fingerprint(headers, path, method)
    return verdict in ("likely_scanner", "likely_ai", "likely_automation")


def _escape_text(s: str) -> str:
    return (
        s.replace("&", "&amp;")
         .replace("<", "&lt;")
         .replace(">", "&gt;")
    )


def _escape_attr(s: str) -> str:
    return (
        s.replace("&", "&amp;")
         .replace('"', "&quot;")
         .replace("<", "&lt;")
         .replace(">", "&gt;")
    )
